cmd/snap/cmd_debug_changes.go

```diff
@@ -0,0 +1,108 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2019 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package main
+
+import (
+	"fmt"
+	"os"
+	"sort"
+	"text/tabwriter"
+	"time"
+
+	"github.com/jessevdk/go-flags"
+	"github.com/snapcore/snapd/i18n"
+	"github.com/snapcore/snapd/overlord/state"
+)
+
+type baseOfflineDebugCommand struct {
+	st *state.State
+
+	Positional struct {
+		StateFilePath string `positional-args:"yes" positional-arg-name:"<state-file>"`
+	} `positional-args:"yes"`
+}
+
+type cmdDebugChanges struct {
+	baseOfflineDebugCommand
+}
+
+var shortDebugChangesHelp = i18n.G("Show all changes from a snapd state file.")
+var longDebugChangesHelp = i18n.G("Show all changes from a snapd state file, bypassing snapd API.")
+
+type byChangeID []*state.Change
+
+func (c byChangeID) Len() int           { return len(c) }
+func (c byChangeID) Swap(i, j int)      { c[i], c[j] = c[j], c[i] }
+func (c byChangeID) Less(i, j int) bool { return c[i].ID() < c[j].ID() }
+
+func formatTime(t time.Time) string {
+	return t.Format(time.RFC3339)
+}
+
+func loadState(path string) (*state.State, error) {
+	if path == "" {
+		path = "state.json"
```

Comment:
Should this case fall back to open the actual `/var/lib/snapd/state.json` file?

Reply:
I'm on the fence with this, it depends on typical use case. The driver for this was to debug someone else's state.json, meaning I'm usually interested in state.json from my CWD.